compiler/rustc_hir_typeck/src/coercion.rs

```diff
@@ -1609,12 +1609,26 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
                 if visitor.ret_exprs.len() > 0 && let Some(expr) = expression {
                     self.note_unreachable_loop_return(&mut err, &expr, &visitor.ret_exprs);
                 }
+
+                if expected.to_string().contains("fn") & found.to_string().contains("fn") {
+                    self.note_function_pointers(&mut err);
+                }
+
                 let reported = err.emit_unless(unsized_return);
 
                 self.final_ty = Some(fcx.tcx.ty_error_with_guaranteed(reported));
             }
         }
     }
+
+    fn note_function_pointers(&self, err: &mut Diagnostic) {
+        err.help(
+            "The item types of different functions - different items, or the same item with different generics \
+            - are distinct, and mixing them will create a type error. \
```

Comment:
"mixing" is an uncommon verb choice here, maybe "can never be assigned to each other"

Reply:
re-worded :)